lib/gtfs_realtime_viz/state.ex

```diff
@@ -14,21 +14,67 @@ defmodule GTFSRealtimeViz.State do
 
   @spec init(:ok) :: {:ok, state}
   def init(:ok) do
-    {:ok, %{}}
+    {:ok, %{vehicles: %{}, trip_updates: %{}}}
   end
 
   # client interface
 
-  @spec new_data(GenServer.server, term, Proto.raw, String.t) :: :ok
-  def new_data(pid \\ __MODULE__, group, raw, comment)
-  def new_data(pid, group, raw, comment) do
-    vehicles =
+  @spec single_pb(GenServer.server, term, Proto.raw, String.t) :: :ok
+  def single_pb(pid \\ __MODULE__, group, raw, comment) do
+    data =
       raw
       |> Proto.FeedMessage.decode
       |> Map.get(:entity)
-      |> Enum.map(& &1.vehicle)
+
+    vehicles = Enum.map(data, & &1.vehicle) |> Enum.reject(& &1 == nil)
+    trip_updates = Enum.map(data, & &1.trip_update) |> Enum.reject(& &1 == nil)
+
+    if !Enum.empty?(vehicles) do
+      new_data(pid, group, raw, nil, comment)
+    end
+    if !Enum.empty?(trip_updates) do
+      new_data(pid, group, nil, raw, comment)
+    end
+  end
+
+  @spec new_data(GenServer.server, term, Proto.raw, Proto.raw, String.t) :: :ok
+  def new_data(pid \\ __MODULE__, group, vehicle_positions, trip_updates, comment)
+  def new_data(pid, group, vehicle_positions, nil, comment) do
+    raw_vp =
+      vehicle_positions
+      |> Proto.FeedMessage.decode
+      |> Map.get(:entity)
+
+    vehicles = Enum.map(raw_vp, & &1.vehicle) |> Enum.reject(& &1 == nil)
+    GenServer.call(pid, {:vehicles, group, vehicles, comment})
```

Comment:
I think we'll need a `:vehicles_and_trip_updates` message. Handling calls within a GenServer is the only way to ensure atomicity of an operation. As it is, this code is running in the process that's calling this function, not the GenServer, and since Erlang can pause execution of a function at any time to let another process have some time, it's possible that this execution gets paused right between these two `GenServer.call`'s, meaning that another process could ask this GenServer its state and see lists of two different lengths.

Reply:
ok, that makes sense.